wsagent/che-core-api-languageserver/src/main/java/org/eclipse/che/api/languageserver/LanguageServerConfig.java

```diff
@@ -0,0 +1,189 @@
+/*
+ * Copyright (c) 2012-2018 Red Hat, Inc.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ *
+ * Contributors:
+ *   Red Hat, Inc. - initial API and implementation
+ */
+package org.eclipse.che.api.languageserver;
+
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.util.Map;
+import java.util.Set;
+import org.eclipse.lsp4j.services.LanguageClient;
+import org.eclipse.lsp4j.services.LanguageServer;
+
+/**
+ * Language server internal configuration that is used by corresponding components for the
+ * definition of language server instance launch, initialization, matching procedures, while it also
+ * specify language server specific configuration elements (like file watch patterns, etc.)
+ *
+ * @see LanguageServerConfigProvider
+ * @see LanguageServerConfigInitializer
+ */
+public interface LanguageServerConfig {
```

Comment:
>Generally, I think this interface is way too complicated, as far as the client is concerned all it needs it is a way to get an instance of a language server and perhaps a way to monitor the status of this instance. The old interface nicely captured this. The fact that you're mandating the use of streams ("CommunicationProvider") in the interface is getting an implementation detail into the interface and unnecessarily restricts the implementation.

The overcomplicating is caused by the need of backward compatibility with the installers, which is expected to be removed in future, then the configuration become simpler, `ComplicationProvider` and all implementation dependent stuff will be removed.

>The fact that we have null implementations and null values (for example the CommunicationProvider) is a code smell that indicates we're describing an implementation, not an interface.

The fact that we have null implementation of `CommunicationProvider` and all similar components is the result of a short-sighted implementation of the [`MavenLanguageServer`](https://github.com/eclipse/che/blob/master/plugins/plugin-maven/che-plugin-maven-server/src/main/java/org/eclipse/che/plugin/maven/lsp/MavenLanguageServer.java) and the necessity to keep backward compatibility, because it is neither an installed language server, nor a parallel container language server.